Mathlib/Topology/Alexandroff.lean

```diff
@@ -0,0 +1,510 @@
+
+instance : Inhabited (Alexandroff X) := ⟨∞⟩
+
+instance [Fintype X] : Fintype (Alexandroff X) :=
+  inferInstanceAs (Fintype (Option X))
+
+instance infinite [Infinite X] : Infinite (Alexandroff X) :=
+  inferInstanceAs (Infinite (Option X))
+#align alexandroff.infinite Alexandroff.infinite
+
+theorem coe_injective : Function.Injective ((↑) : X → Alexandroff X) :=
+  Option.some_injective X
+#align alexandroff.coe_injective Alexandroff.coe_injective
+
+@[norm_cast]
+theorem coe_eq_coe {x y : X} : (x : Alexandroff X) = y ↔ x = y :=
+  coe_injective.eq_iff
+#align alexandroff.coe_eq_coe Alexandroff.coe_eq_coe
+
+@[simp]
+theorem coe_ne_infty (x : X) : (x : Alexandroff X) ≠ ∞ :=
+  fun.
+#align alexandroff.coe_ne_infty Alexandroff.coe_ne_infty
+
+@[simp]
+theorem infty_ne_coe (x : X) : ∞ ≠ (x : Alexandroff X) :=
+  fun.
+#align alexandroff.infty_ne_coe Alexandroff.infty_ne_coe
+
+/-- Recursor for `Alexandroff` using the preferred forms `∞` and `↑x`. -/
+@[elab_as_elim]
+protected def rec {C : Alexandroff X → Sort _} (h₁ : C ∞) (h₂ : ∀ x : X, C x) :
+    ∀ z : Alexandroff X, C z
+  | ∞ => h₁
+  | (x : X) => h₂ x
+#align alexandroff.rec Alexandroff.rec
+
+theorem isCompl_range_coe_infty : IsCompl (range ((↑) : X → Alexandroff X)) {∞} :=
+  isCompl_range_some_none X
+#align alexandroff.is_compl_range_coe_infty Alexandroff.isCompl_range_coe_infty
+
+-- porting note: moved @[simp] to a new lwmma
+theorem range_coe_union_infty : range ((↑) : X → Alexandroff X) ∪ {∞} = univ :=
+  range_some_union_none X
+#align alexandroff.range_coe_union_infty Alexandroff.range_coe_union_infty
+
+@[simp]
+theorem insert_infty_range_coe : insert ∞ (range (@some X)) = univ :=
+  insert_none_range_some _
+
+@[simp]
+theorem range_coe_inter_infty : range ((↑) : X → Alexandroff X) ∩ {∞} = ∅ :=
+  range_some_inter_none X
+#align alexandroff.range_coe_inter_infty Alexandroff.range_coe_inter_infty
+
+@[simp]
+theorem compl_range_coe : range ((↑) : X → Alexandroff X)ᶜ = {∞} :=
+  compl_range_some X
+#align alexandroff.compl_range_coe Alexandroff.compl_range_coe
+
+theorem compl_infty : ({∞}ᶜ : Set (Alexandroff X)) = range ((↑) : X → Alexandroff X) :=
+  (@isCompl_range_coe_infty X).symm.compl_eq
+#align alexandroff.compl_infty Alexandroff.compl_infty
+
+theorem compl_image_coe (s : Set X) : ((↑) '' s : Set (Alexandroff X))ᶜ = (↑) '' sᶜ ∪ {∞} := by
+  rw [coe_injective.compl_image_eq, compl_range_coe]
+#align alexandroff.compl_image_coe Alexandroff.compl_image_coe
+
+theorem ne_infty_iff_exists {x : Alexandroff X} : x ≠ ∞ ↔ ∃ y : X, (y : Alexandroff X) = x := by
+  induction x using Alexandroff.rec <;> simp
+#align alexandroff.ne_infty_iff_exists Alexandroff.ne_infty_iff_exists
+
+instance canLift : CanLift (Alexandroff X) X (↑) fun x => x ≠ ∞ :=
+  WithTop.canLift
+#align alexandroff.can_lift Alexandroff.canLift
+
+theorem not_mem_range_coe_iff {x : Alexandroff X} : x ∉ range ((↑) : X → Alexandroff X) ↔ x = ∞ :=
+  by rw [← mem_compl_iff, compl_range_coe, mem_singleton_iff]
```

Comment:
Is that `by` at the beginning of a line something we accept in the new convention? (I don't know)
If not:
```suggestion
theorem not_mem_range_coe_iff {x : Alexandroff X} :
    x ∉ range ((↑) : X → Alexandroff X) ↔ x = ∞ := by
  rw [← mem_compl_iff, compl_range_coe, mem_singleton_iff]
```

Reply:
I used `some` to save a line.